mrmustard/lab/states/dm.py

```diff
@@ -306,7 +306,28 @@ def expectation(self, operator: CircuitComponent, mode: str = "kron") -> Batch[S
         elif op_type is OperatorType.DM_LIKE:
             result = self.contract(operator.dual, mode=mode) >> TraceOut(leftover_modes)
         else:
-            result = (self.contract(operator, mode=mode)) >> TraceOut(self.modes)
+            # custom shape handling from contract
+            # since input and output wires are contracted
+            # to do the trace out
+            if (
+                type(self.ansatz) is not type(operator.ansatz)
+                and settings.DEFAULT_REPRESENTATION == "Fock"
+            ):
+                self_shape = list(self.auto_shape())
+                other_shape = list(operator.auto_shape())
+                # want to make sure that only the operator modes use shape lookahead
+                # for efficiency
+                for m in operator.modes:
+                    for idx1, idx2 in zip(self.wires[m].indices, operator.wires[m].indices):
```

Comment:
Is the `zip` here a "trick" to loop over the iterable with the fewest items? You don't _need_ two indices, do you @apchytr?

Reply:
Actually, does this work properly for different-sized tuples in the `zip`?

Maybe only a problem for `Ket`s and not `DM`s...